### ISSUE TITLE
RuntimeError when monkey patching gevent

### ISSUE THREAD

**Issue by dhhagan** [miguelgrinberg/Flask-SocketIO#901]
After adding a message queue and trying to monkey patch gevent on the flask-socketio dev server, I began to see a RuntimeError:

<pre>
/Users/davidhagan/Documents/github/aerovir/ceams/ceams/manage.py:2: MonkeyPatchWarning: Monkey-patching outside the main native th
read. Some APIs will not be available. Expect a KeyError to be printed at shutdown.
  monkey.patch_all(subprocess=True)
/Users/davidhagan/Documents/github/aerovir/ceams/ceams/manage.py:2: MonkeyPatchWarning: Monkey-patching not on the main thread; th
reading.main_thread().join() will hang from a greenlet
  monkey.patch_all(subprocess=True)

Exception in thread Thread-1:
Traceback (most recent call last):
  File "/Users/davidhagan/.pyenv/versions/3.6.5/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/Users/davidhagan/.pyenv/versions/3.6.5/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/davidhagan/.pyenv/versions/3.6.5/envs/ceams_env/lib/python3.6/site-packages/flask_socketio/cli.py", line 53, in run
_server
    app = info.load_app()
  File "/Users/davidhagan/.pyenv/versions/3.6.5/envs/ceams_env/lib/python3.6/site-packages/flask/cli.py", line 372, in load_app
    app = locate_app(self, import_name, name)
  File "/Users/davidhagan/.pyenv/versions/3.6.5/envs/ceams_env/lib/python3.6/site-packages/flask/cli.py", line 235, in locate_app
    __import__(module_name)
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 152, in __exit__
  File "<frozen importlib._bootstrap>", line 107, in release
RuntimeError: cannot release un-acquired lock
</pre>

I've never seen this before and it didn't start until I tried to monkey patch so that I could (hopefully) use socketio.emit from an RQ background task.

Environment:

<pre>
Flask==1.0.2
Flask-SocketIO==3.1.2
gevent==1.4.0
gevent-websocket==0.10.1
greenlet==0.4.15
gunicorn==19.9.0
...
</pre>

### `manage.py`

<pre>
from gevent import monkey
monkey.patch_all(subprocess=True)

# load the environment if it exists
dotenv_path = os.path.join(os.path.dirname(__name__), ".flaskenv")
if os.path.exists(dotenv_path):
    load_dotenv(dotenv_path)

# set up coverage for unittesting
COV = None
if os.environ.get("FLASK_COVERAGE"):
    import coverage
    COV = coverage.coverage(branch=True, include="app/*")
    COV.start()

import sys
import click
from flask_migrate import Migrate, upgrade
from app import create_app, db
import sqlalchemy
from app.models import User, Role, DeploymentTag, Post, Settings, Notification

app = create_app(os.getenv("FLASK_CONFIG") or 'default')
...

</pre>

### `app/__init__.py`

<pre>
from flask import Flask, request, current_app
from flask_socketio import SocketIO
from flask_login import LoginManager
from flaskext.markdown import Markdown
from flask_sqlalchemy import SQLAlchemy
from flask_pagedown import PageDown
from flask_moment import Moment
from flask_caching import Cache
from flask_apscheduler import APScheduler
from flask_rq2 import RQ

from flask_assets import Environment, Bundle
from config import config
from redis import Redis
import rq

db = SQLAlchemy()
socketio = SocketIO()
login = LoginManager()
assets = Environment()
pagedown = PageDown()
moment = Moment()
cache = Cache(config={"CACHE_TYPE": "simple"})
rq_app = RQ()
scheduler = APScheduler()

login.login_view = "auth.login"

def create_app(config_name):
    app = Flask(__name__)
    app.config.from_object(config[config_name])
    config[config_name].init_app(app)

    ...

    # init things
    db.init_app(app)
    login.init_app(app)
    socketio.init_app(app, message_queue=app.config.get("REDIS_URL"))
    assets.init_app(app)
    pagedown.init_app(app)
    moment.init_app(app)
    cache.init_app(app)
    rq_app.init_app(app)

    if app.debug:
        scheduler.init_app(app)
        scheduler.start()

    Markdown(app)

    # register blueprints
    from app.auth import bp as auth_bp
    from app.main import bp as main_bp
    from app.admin import admin as admin_bp
    from app.api_v1 import api as api_v1_bp
    from app.errors import bp as errors_bp

    app.register_blueprint(main_bp)
    app.register_blueprint(auth_bp, url_prefix="/auth")
    app.register_blueprint(api_v1_bp, url_prefix="/api/v1")
    app.register_blueprint(admin_bp, url_prefix="/admin")
    app.register_blueprint(errors_bp)

    return app

from app import models
</pre>

**Comment by miguelgrinberg** [comment]
How do you start your server?

**Comment by dhhagan** [comment]
In this case, it's just the dev server, so `flask run` (also updated above to include a bit more of `manage.py`). In production it's a gunicorn setup, but haven't gotten that far with this config.

**Comment by miguelgrinberg** [comment]
Try using `flask run --eager-loading`. Does that eliminate the monkey patching warnings during start?

**Comment by dhhagan** [comment]
Unfortunately it does not...

**Comment by miguelgrinberg** [comment]
So you still get the warning about the monkey patching being done in a thread?

Instead of using `flask run`, add

    if __name__ == '__main__':
        app.run()

and then start the server by running `python manage.py`. Does that get rid of the warnings?

**Comment by dhhagan** [comment]
Yes, that does get rid of the warning!

**Comment by miguelgrinberg** [comment]
Okay. I stopped supporting the `flask run` command for Socket.IO in the last release, because there are a lot of little issues like this one that are hard to address because the `run` command cannot be easily modified. You are not using the most current release and for that reason the package is still "trying" to make things work.

So I recommend you start the application by running the script. I just noticed that I gave you an incorrect start up line for Socket.IO, by the way, it should have been:

```python
if __name__ == '__main__':
    socketio.run(app)
```

**Comment by insensitiveclod** [comment]
Wow..
I got hit by exactly this issue with "flask run" bombing out with messages like this:

`
[2019-02-12 14:00:06,197] WARNING in __init__: Flask-SocketIO is Running under Werkzeug, WebSocket is not available.
 * Running on http://0.0.0.0:5004/ (Press CTRL+C to quit)
Traceback (most recent call last):
...
...
  self._poll = select.poll()
AttributeError: module 'select' has no attribute 'poll'
`
It took me a couple of hours to pin down what the issue was with a new project I was working on. Only when i upgraded the venv of some old (working) projects did I find out it was due to Flask-SocketIO version-differences.

The only hint I had was that there was *some* reason why Flask-SocketIO wasnt using Websockets. No clue as to why.

Of note is also that most/all flask-tutorials refer to 'flask run' as the method to run an application.

Is there either:
- A way to add code to a project to make 'flask run' work once more as it used to ?
- Add (more) output at the top as to WHY websocket-support isnt enabled/available and how to fix that ?
